Feedback improvements

kspurgin opened the pull request.

kspurgin commented:
@mark-cooper I think I have this working. At least a first version. Apologies that it is so big and messy, but the feedback was already woven through a bunch of files and tests. 

I think this overlaps with and addresses what's in PR #97 to some extent, because I was getting confused about what was actually not working due to changes vs. what was related to flaky tests.

mark-cooper commented:
@kspurgin I'm going to be awkward on this one and suggest we first merge: https://github.com/dts-hosting/data-toolkit/pull/97

I think that PR provides a good starting foundation for how to approach running and asserting with the jobs in tests. In particular I want to avoid a few things like:

https://github.com/dts-hosting/data-toolkit/pull/73/files#diff-96cf2e592460bc45d1db7680e0b1399e1c742cfea1050e9426389a13e90fb16bR7-R24

The unscoped use of  `perform_enqueued_jobs # item jobs, should all succeed` with comments to describe what is happening and direct calls to `task.update_progress`.

As is this is also hard to follow: https://github.com/dts-hosting/data-toolkit/pull/73/files#diff-96cf2e592460bc45d1db7680e0b1399e1c742cfea1050e9426389a13e90fb16bR8-R11

Just because we need a more expressive, encapsulated way of resetting task state with items, which is in #97 (though can be improved further when / if some patterns emerge).

And same with something like: https://github.com/dts-hosting/data-toolkit/pull/73/files#diff-96cf2e592460bc45d1db7680e0b1399e1c742cfea1050e9426389a13e90fb16bR97-R106

Obviously merging #97 will create a merge conflict. Because I'm the one being awkward I'm fine with taking on the task of pushing a cleanup commit. Unless you want to do it for the all the fun and joy it brings to do conflict resolution in git? 😄

mark-cooper commented:
I'd like to consolidate these into a single file (`app/exceptions/exceptions.rb`?):

- app/exceptions/data_toolkit_error.rb
- app/exceptions/feedback_subtype_error.rb

Over time we could end up with a bunch of:

```ruby
class SomeCustomError < StandardError
class SomeOtherCustomError < StandardError
```

I think it's totally fine to put all of those into a single file.

Overall this is looking good, I'll try it out later today!

kspurgin commented:
I went back and forth on that. I have [one errors.rb file in collectionspace-mapper](https://github.com/collectionspace/collectionspace-mapper/blob/main/lib/collectionspace/mapper/error.rb), like you recommend, but saw conflicting recommendations about how to do it in Rails. I thought determined the "one file per class" was winning out, due to Zeitwerk assumptions, but if one file works, I do actually prefer that for teeny little similar classes like exceptions.

kspurgin commented:
@mark-cooper - I should admit I didn't actually get around to testing whether the views actually work. I got it to an all-tests passing, working on the back end, state later than I should have been working last night. Figured I should make sure you didn't have big-change feedback before spending time ensuring display was good.

kspurgin commented:
@mark-cooper Hey, it's not awkward to assert technical preferences/opinions as the senior tech professional on a project!

I'll merge the other and deal with any conflicts here.

kspurgin commented:
@mark-cooper Conflicts are resolved in this

mark-cooper commented:
https://github.com/dts-hosting/data-toolkit/pull/73/files#diff-4095b65754098dce7229824782d43e80a7cb3f0fe45aa245102593760b51a24bR94-R95

Re:

```rb
data_items.where(status: "failed").exists? ? fail!(feedback) : success!
```

Conceptually this looks odd because we're passing in the current state of feedback to be re-saved (which I imagine looks something like `UPDATE ... SET $feedback = $feedback`, which would be redundant and probably should be avoided (so the optional {} arg could be reintroduced and used to only update feedback if not empty and don't pass it when it would be updating with no change). Practically speaking this may be minor if feedback is mostly provided after this update by a finalizer, so I'll defer to you on whether the change is worth it at this stage.

---

https://github.com/dts-hosting/data-toolkit/pull/73/files#diff-4535946f385c5e3fe3581beb35863c41575d12b10410b7569c54037d4e2a5058R22
https://github.com/dts-hosting/data-toolkit/pull/73/files#diff-8da07ea812d023fda049fe0c5deed90dd579832d7f206728a4171f9fdc4dedf5R19

Re:

```rb
@feedback = Feedback.new("Tasks::ProcessUploadedFiles")
```

I'm not a DI purist but I think it is structurally better to pass feedback through because it's already set for `FileValidator` in the "expected" way.

Roughly sketched that's something like:

```diff
diff --git a/app/services/files_validator.rb b/app/services/files_validator.rb
index b56022d..cbbdc00 100644
--- a/app/services/files_validator.rb
+++ b/app/services/files_validator.rb
@@ -14,7 +14,7 @@ class FilesValidator

   def call
     begin
-      @results = files.map { |file| pick_validator(file).call(file) }
+      @results = files.map { |file| pick_validator(file).call(file, @feedback) }
     rescue => err
       @validity_status = false
       Rails.logger.error "#{err.message} -- #{err.backtrace.first(5)}"
```

```diff
diff --git a/app/services/csv_validator.rb b/app/services/csv_validator.rb
index 1712df1..383ff83 100644
--- a/app/services/csv_validator.rb
+++ b/app/services/csv_validator.rb
@@ -5,8 +5,9 @@ class CsvValidator

   # @param file [ActiveStorage::Attachment]
   # @param dialect [Hash]
-  def initialize(file, dialect: DEFAULT_DIALECT)
+  def initialize(file, feedback, dialect: DEFAULT_DIALECT)
     @file = file
+    @feedback = feedback
     @dialect = dialect
   end

@@ -14,9 +15,9 @@ class CsvValidator
     @filename = file.filename.to_s
     file.open do |f|
       path = Pathname.new(f.path)
-      @csvlint = CsvlintValidator.call(path, filename: filename,
+      @csvlint = CsvlintValidator.call(path, @feedback, filename: filename,
         dialect: dialect)
-      @stdlib = CsvStdlibValidator.call(path, filename: filename)
+      @stdlib = CsvStdlibValidator.call(path, @feedback, filename: filename)
     end
```

```diff
diff --git a/app/services/csv_stdlib_validator.rb b/app/services/csv_stdlib_validator.rb
index 58f4290..a60c4a2 100644
--- a/app/services/csv_stdlib_validator.rb
+++ b/app/services/csv_stdlib_validator.rb
@@ -16,10 +16,10 @@ class CsvStdlibValidator
   # @param filename [nil, String] the user-known filename is not the basename
   #   of the Rails storage path, so this should be passed explicitly in
   #   production, so the feedback can record which file it is about
-  def initialize(filepath, filename: nil)
+  def initialize(filepath, feedback, filename: nil)
     @filepath = filepath
     @filename = filename || filepath.basename.to_s
-    @feedback = Feedback.new("Tasks::ProcessUploadedFiles")
+    @feedback = feedback
     @feedback_subtype_prefix = :csv_stdlib
   end

diff --git a/app/services/csvlint_validator.rb b/app/services/csvlint_validator.rb
index 9696d7d..177f29e 100644
--- a/app/services/csvlint_validator.rb
+++ b/app/services/csvlint_validator.rb
@@ -11,12 +11,12 @@ class CsvlintValidator
   #   of the Rails storage path, so this should be passed explicitly in
   #   production, so the feedback can record which file it is about
   # @param dialect [Hash]
-  def initialize(filepath, filename: nil,
+  def initialize(filepath, feeback, filename: nil,
     dialect: CsvValidator::DEFAULT_DIALECT)
     @filepath = filepath
     @filename = filename || filepath.basename.to_s
     @dialect = dialect
-    @feedback = Feedback.new("Tasks::ProcessUploadedFiles")
+    @feedback = feeback
     @feedback_subtype_prefix = :csvlint
   end
```

The idea is just to be consistent in passing the feedback through and not have "hidden" places that hard code a dependency. If it needs to be a new instance of feedback for each validator then can just pass the class.

Looking at the validators there's consistent use of `def self.call(...) = new(...).call` which looks a little too indirect to be my preference (I find myself jumping around a bit to see what's going on), and a bit harder to follow if you're not familiar with several Ruby concepts (class vs instance methods for `call` used in conjunction with range args for the class method and constructor) but that's a stylistic thing and if you like it that's totally fine.

---

General observation, a lot of work goes into properly validating some CSV files for CSpace!

---

A small thing in the feedback view:

```diff
--- a/app/views/tasks/_feedback.html.erb
+++ b/app/views/tasks/_feedback.html.erb
@@ -6,7 +6,7 @@
   <% feedback.each do |category, entries| %>
     <div class="card-body">
       <div class="mb-4">
-        <h6 class="text-muted fw-semibold border-bottom pb-2 mb-3"><%= category.titleize %></h6>
+        <h6 class="text-muted fw-semibold border-bottom pb-2 mb-3"><%= category.to_s.titleize %></h6>
```

This might be wrapped in a helper at some point ... but for now this is fine.

---

I'll take a look at the tests later today!

mark-cooper commented:
Tests look good! Just spotted one place where it would be good to wrap a perform enqueue:

```diff
diff --git a/test/jobs/pre_check_ingest_data_job_test.rb b/test/jobs/pre_check_ingest_data_job_test.rb
index e196b23..998eaf0 100644
--- a/test/jobs/pre_check_ingest_data_job_test.rb
+++ b/test/jobs/pre_check_ingest_data_job_test.rb
@@ -24,7 +24,11 @@ class PreCheckIngestDataJobTest < ActiveJob::TestCase

     assert @task.ok_to_run?
     @task.run
-    perform_enqueued_jobs
+
+    assert_performed_with(job: PreCheckIngestDataJob, args: [@task]) do
+      perform_enqueued_jobs
+    end
+
     @task.reload
     assert_equal "failed", @task.status
```

kspurgin commented:
@mark-cooper - I have added [a commit](https://github.com/dts-hosting/data-toolkit/pull/73/commits/ee1136851908552ef53ece70c4677135a4b68d60) with tests that indicate to me that feedback is working as I expected when passed through the `:fail!` method.

I don't feel like `task_tests.rb` is really the correct place for all of those, but also don't know where a better place would be.

The ":feedback_for should raise exception when invalid feedback Hash has been saved" test is intended to be illustrative for the moment and not defining ultimately desired behavior. 

Although, nobody should be writing code in this app that passes random content-bearing Hashes as feedback---that was the whole point of implementing the Feedback stuff in this PR---so 🤷🏻.

As to that AND the Task.finish_up method doing `fail!(feedback)` instead of `fail!({})`... I wish to defer that change because I think it needs to be addressed in #102. 

If you make that change in a local copy and run the tests, `PreCheckIngestDataFinalizerJobTest` "fails task and adds error to feedback if any item jobs failed" test will fail with the same error seen in the `TaskTest` ":feedback_for should raise exception when invalid feedback Hash has been saved" test I added in this commit. 

The test that fails if you make this change is [one of the ones you changed the setup for](https://github.com/dts-hosting/data-toolkit/commit/1d5854a549f23c9a03daeef060ecfb58bc9445c8). I'm still not 100% on why this new setup causes the test to fail. The tests I added in the commit on this PR show that, with the expected usage, doing `fail!({})` is handled. I am not seeing where any partial/invalid feedback hashes are being set anywhere in the setup. 

But this test is directly related to the Task.finish_up and finalizer run relationship, which I feel needs to be fixed, which will also fix whatever this issue is. I think.

kspurgin commented:
AND I'll fix that double iteration in that code block I just pasted in... we don't need to map then inject. Sigh.

mark-cooper commented:
@kspurgin I left this comment: https://github.com/dts-hosting/data-toolkit/issues/102#issuecomment-3015527825

Re: how to handle the status update with feedback:

```ruby
# from
data_items.where(status: "failed").exists? ? fail!(feedback) : success!
# to
data_items.where(status: "failed").exists? ? fail! : success!
# finalizer is responsible for compiling feedback
```

And the `TransitionsStatus` concern doesn't do anything with feedback.

kspurgin commented:
@mark-cooper - I believe all your comments have been addressed.

mark-cooper commented:
Looking good so far! Thx. I want to step through it some more but I'll aim to review all of it asap so we can keep moving forward.

mark-cooper commented:
So with this PR we have to pick a way forward on whether a finalizer has any business setting status.

If we go with the plan: https://github.com/dts-hosting/data-toolkit/issues/112

```diff
diff --git a/app/jobs/pre_check_ingest_data_finalizer_job.rb b/app/jobs/pre_check_ingest_data_finalizer_job.rb
index fa59f0f..901acda 100644
--- a/app/jobs/pre_check_ingest_data_finalizer_job.rb
+++ b/app/jobs/pre_check_ingest_data_finalizer_job.rb
@@ -8,11 +8,11 @@ class PreCheckIngestDataFinalizerJob < ApplicationJob

     log_finish && return if item_failures.empty?

-    task.fail!(item_failure_feedback_for(feedback, item_failures))
+    task.update(feedback: item_failure_feedback_for(feedback, item_failures))
   rescue => e
     Rails.logger.error e.message
     feedback.add_to_errors(subtype: :application_error, details: e)
-    task.fail!(feedback)
+    task.update(feedback: feedback)
   end

   private
```

The finalizer only updates feedback and can add files. Status has already been determined.

Looking at it from the current PR perspective with the finalizer actually updating task status we could do this:

```diff
diff --git a/app/models/task.rb b/app/models/task.rb
index 558ead9..685e518 100644
--- a/app/models/task.rb
+++ b/app/models/task.rb
@@ -92,7 +92,6 @@ class Task < ApplicationRecord
   end

   def finish_up
-    data_items.where(status: "failed").exists? ? fail! : success!
     finalizer&.perform_later(self)
   end
```

Now the responsibility for setting the status is with the finalizer (which is what `pre_check_ingest_data_finalizer_job` is doing in the PR).

My preference remains with the former option. I think it's simpler (there's a "universal" way of understanding how any / all task status is determined), and the status is (or can be known) before feedback and reports are added (and an error running the finalizer itself isn't a task concern per se). But I'm flexible, and I'm open to the idea that there's a good reason to defer setting the status to the finalizer.

I do very much do want to avoid per task implementations of methods like `finish_up` so if setting status in the finalizer works without adding other deviations I can be more in favor of it.

@kspurgin you're very likely going to be implementing the majority of these jobs. Do you want to move forward with the finalizer setting status like in this PR?

In any case I think we should consider whether we can simplify some things by adding a "warning" or "review" status to task and data_item (> 0 fails = fail, else > 0 warnings/review = warning/review, else success [at the task level]).

Crudely that would enable something like:

```diff
diff --git a/app/jobs/pre_check_ingest_data_finalizer_job.rb b/app/jobs/pre_check_ingest_data_finalizer_job.rb
index fa59f0f..3531bff 100644
--- a/app/jobs/pre_check_ingest_data_finalizer_job.rb
+++ b/app/jobs/pre_check_ingest_data_finalizer_job.rb
@@ -5,10 +5,18 @@ class PreCheckIngestDataFinalizerJob < ApplicationJob
     Rails.logger.info "#{self.class.name} started"
     feedback = task.feedback_for
     item_failures = task.data_items.where(status: "failed")
+    items_to_review = task.data_items.where(status: "review")

-    log_finish && return if item_failures.empty?
+    log_finish && return if item_failures.empty? && items_to_review.empty?

-    task.fail!(item_failure_feedback_for(feedback, item_failures))
+    feedback = item_feedback_for(feedback, item_failures, items_to_review)
+    if item_failures.any?
+      task.fail!(feedback)
+    elsif items_to_review.any?
+      task.review!(feedback)
+    else
+      task.success!
+    end
   rescue => e
     Rails.logger.error e.message
     feedback.add_to_errors(subtype: :application_error, details: e)
@@ -21,12 +29,17 @@ class PreCheckIngestDataFinalizerJob < ApplicationJob
     Rails.logger.info "#{self.class.name} finished"
   end

-  def item_failure_feedback_for(feedback, failures)
+  def item_feedback_for(feedback, failures, review)
     failures.each do |item|
       item.feedback_for.errors.each do |err|
         feedback.add_to_errors(subtype: err.subtype, details: err.details)
       end
     end
+    review.each do |item|
+      item.feedback_for.warnings.each do |err|
+        feedback.add_to_warnings(subtype: err.subtype, details: err.details)
+      end
+    end
     feedback
   end
 end
```
